<a href="https://colab.research.google.com/github/OsvaldoFloresC/OsvaldoFloresC/blob/main/Polinomio_de_Hermite.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# **Polinomio de Hermite**
La interpolación de Hermite es una técnica avanzada de interpolación polinómica que no solo se basa en los valores de una función en ciertos puntos (como en la interpolación de Lagrange), sino que también utiliza las derivadas de la función en esos puntos. Este método permite construir un polinomio que no solo pasa por los puntos dados, sino que también coincide con las pendientes de la función en esos puntos, proporcionando una aproximación más precisa y suave.
$$P_n(x) = f[x_0] + \sum_{k=1}^{n} f[x_0, x_1, \ldots, x_k] \prod_{j=0}^{k-1} (x - x_j)$$
Donde\
  $ f[x_0] $ es el valor de la función en $( x_0 )$ \
  $ f[x_0, x_1, \ldots, x_k] $ son las diferencias divididas de la función. \
  $\prod_{j=0}^{k-1} (x - x_j)$ es el producto de los términos $(x - x_j)$ desde $ j = 0 $ hasta $j = k-1 $.

In [ ]:
import numpy as np
import sympy as sp

def metodo_hermite(x, xi, fi, fpi):
    # Número de puntos
    n = len(xi)

    # Crear una tabla vacía para los cálculos (2n x 2n)
    Q = np.zeros((2 * n, 2 * n))
    z = np.zeros(2 * n)

    # Poner los valores iniciales (f(xi)) y (f'(xi)) en la tabla
    for i in range(n):
        z[2 * i] = z[2 * i + 1] = xi[i]
        Q[2 * i][0] = Q[2 * i + 1][0] = fi[i]
        Q[2 * i + 1][1] = fpi[i]
        if i != 0:
            Q[2 * i][1] = (Q[2 * i][0] - Q[2 * i - 1][0]) / (z[2 * i] - z[2 * i - 1])

    # Llenar la tabla usando la fórmula del método de Hermite
    for i in range(2, 2 * n):
        for j in range(2, i + 1):
            Q[i][j] = (Q[i][j - 1] - Q[i - 1][j - 1]) / (z[i] - z[i - j])

    # Redondear los resultados a 7 cifras decimales
    Q = np.round(Q, 7)

    # Coeficientes del polinomio de Hermite
    coeficientes = [Q[i][i] for i in range(2 * n)]

    # Construir el polinomio de Hermite
    polinomio = f"{coeficientes[0]}"
    for i in range(1, 2 * n):
        term = f"{coeficientes[i]}"
        for j in range(i):
            term += f"*(x - {z[j]})"
        polinomio += " + " + term

    # Calcular el valor del polinomio en x
    valor_polinomio = coeficientes[0]
    for i in range(1, 2 * n):
        term_value = coeficientes[i]
        for j in range(i):
            term_value *= (x - z[j])
        valor_polinomio += term_value

    # Crear tabla con encabezados
    tabla_hermite = [["xi", "f(x)", "PDD", "SDD", "TDD"]]
    for i in range(2 * n):
        fila = [z[i], Q[i][0]]
        for j in range(1, i + 1):
            if j < len(tabla_hermite[0]):
                fila.append(Q[i][j])
            else:
                fila.append("")
        tabla_hermite.append(fila)

    return valor_polinomio, Q, polinomio, tabla_hermite

# Valores iniciales para la tabla de Hermite
xi = [1.3, 1.6, 1.9]  # Valores de x
fi = [0.6200860, 0.4554022, 0.2818186]  # Valores de f(x)
fpi = [-0.5220232, -0.5698959, -0.5811571]  # Valores de f'(x)
x = 1.5  # El valor donde queremos interpolar

# Calcular el polinomio y la tabla de Hermite
valor_polinomio, Q, polinomio, tabla_hermite = metodo_hermite(x, xi, fi, fpi)

# Mostrar los resultados
print("Tabla de Hermite:")
for fila in tabla_hermite:
    print(fila)
print(f"H(x) = {polinomio}")

# Verificar el valor en x = 1.5
value = 1.5
a = 1
h = Q[0][0]
for i in range(1, 2 * len(xi)):
    a = a * (value - z[i - 1])
    h = h + a * Q[i][i]

print(f"El valor aproximado en H({x}) = {valor_polinomio}")


Tabla de Hermite:
['xi', 'f(x)', 'PDD', 'SDD', 'TDD']
[1.3, 0.620086]
[1.3, 0.620086, -0.5220232]
[1.6, 0.4554022, -0.548946, -0.0897427]
[1.6, 0.4554022, -0.5698959, -0.069833, 0.0663656]
[1.9, 0.2818186, -0.578612, -0.0290537, 0.0679656, 0.0026667]
[1.9, 0.2818186, -0.5811571, -0.0084837, 0.0685667, 0.0010019, '']
H(x) = 0.620086 + -0.5220232*(x - 1.3) + -0.0897427*(x - 1.3)*(x - 1.3) + 0.0663656*(x - 1.3)*(x - 1.3)*(x - 1.6) + 0.0026667*(x - 1.3)*(x - 1.3)*(x - 1.6)*(x - 1.6) + -0.0027747*(x - 1.3)*(x - 1.3)*(x - 1.6)*(x - 1.6)*(x - 1.9)
El valor aproximado en H(1.5) = 0.5118277002320001


Por lo que el polinomio de Hermite queda de la siguiente manera:

$$H(x) = 0.620086 + -0.5220232 \cdot (x - 1.3) + -0.08974267 \cdot (x - 1.3)^2 + 0.06636556 \cdot (x - 1.3)^2 \cdot (x - 1.6) + 0.00266667 \cdot (x - 1.3)^2 \cdot (x - 1.6)^2 + -0.00277469 \cdot (x - 1.3)^2 \cdot (x - 1.6)^2 \cdot (x - 1.9)$$

Utilizando la libreria SYMPY podemos hacer uso de herramientas de algebra computacional con lo cual el polinomio desarrollado tiene la siguiente forma:


In [ ]:
sp.simplify(polinomio)

-0.0027747*x**5 + 0.02403189*x**4 - 0.014556433*x**3 - 0.2352156601*x**2 - 0.00822951440000003*x + 1.001944103432

Aunque la verdad este ultimo paso no es muy conveniente pues al seguir desarrollando el polinomio el error se propaga, por lo que es mas conveniente trabajar con su forma sin desarrollar.
Este método es especialmente útil cuando se requiere que no solo los valores de la función sean exactos en ciertos puntos, sino también las pendientes (derivadas).